# Data Science for Social Justice Workshop Group Project: Ukraine


## 1. Preprocessing

### 1-1. Importing Data with pandas

In [1]:
%pwd ##it will be different for all

'/home/jovyan/Data-Science-Social-Justice'

In [2]:
import os

In [3]:
os.chdir('data') ##it will be different for all

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv('submissions.csv')

In [6]:
df.shape

(88764, 18)

In [7]:
# 18 variables
list (df)

['idint',
 'idstr',
 'created',
 'self',
 'nsfw',
 'author',
 'title',
 'url',
 'selftext',
 'score',
 'subreddit',
 'distinguish',
 'textlen',
 'num_comments',
 'flair_text',
 'flair_css_class',
 'augmented_at',
 'augmented_count']

In [8]:
df.head()

idint     idstr     created  self  nsfw       author  \
0  13054757  t3_7rt45  1232688225     0     0  nikitos2009   
1  13072291  t3_7s6n7  1232834472     1     0        OlehM   
2  13093545  t3_7sn1l  1233019738     0     0  nikitos2009   
3  13118896  t3_7t6ls  1233182615     0     0  nikitos2009   
4  13238506  t3_7vqwa  1234081946     0     0   ger4antche   

                                               title  \
0               Лучший политический ресурс в Украине   
1                                                 Dr   
2           Основы гармонии украинства с московством   
3  Служба Бандеризации Украины выясняет кто убил ...   
4                                 amzek.blogspot.com   

                                 url   selftext  score subreddit distinguish  \
0               http://pepper.at.ua/        NaN      0   ukraine         NaN   
1                                NaN  [removed]      1   ukraine         NaN   
2  http://pepper.at.ua/publ/8-1-0-30        NaN      1   ukraine         NaN   
3  http://pepper.at.ua/publ/8-1-0-34        NaN      1   ukraine         NaN   
4          http://amzek.blogspot.com        NaN      0   ukraine         NaN   

   textlen  num_comments flair_text flair_css_class  augmented_at  \
0        0             0        NaN             NaN           NaN   
1        9             0        NaN             NaN           NaN   
2        0             0        NaN             NaN           NaN   
3        0             0        NaN             NaN           NaN   
4        0             1        NaN             NaN           NaN   

   augmented_count  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN

### 1-2. Dropping Columns and Missing Values

In [9]:
##remove some columns that we are not going to use

df = df.drop(['self', 'url', 'subreddit', 'augmented_at', 'augmented_count'], axis=1)
df.head()

idint     idstr     created  nsfw       author  \
0  13054757  t3_7rt45  1232688225     0  nikitos2009   
1  13072291  t3_7s6n7  1232834472     0        OlehM   
2  13093545  t3_7sn1l  1233019738     0  nikitos2009   
3  13118896  t3_7t6ls  1233182615     0  nikitos2009   
4  13238506  t3_7vqwa  1234081946     0   ger4antche   

                                               title   selftext  score  \
0               Лучший политический ресурс в Украине        NaN      0   
1                                                 Dr  [removed]      1   
2           Основы гармонии украинства с московством        NaN      1   
3  Служба Бандеризации Украины выясняет кто убил ...        NaN      1   
4                                 amzek.blogspot.com        NaN      0   

  distinguish  textlen  num_comments flair_text flair_css_class  
0         NaN        0             0        NaN             NaN  
1         NaN        9             0        NaN             NaN  
2         NaN        0             0        NaN             NaN  
3         NaN        0             0        NaN             NaN  
4         NaN        0             1        NaN             NaN

In [10]:
##get rid of posts that have been deleted/removed

df = df.loc[~df['selftext'].isin(['[removed]', '[deleted]' ]),:]
df.shape

(73414, 13)

In [11]:
## drop null values
df = df.dropna(subset=['selftext'])
df.shape

(16665, 13)

In [12]:
df.head()

idint     idstr     created  nsfw         author  \
35  26781072  t3_fy0eo  1299351508     0         margys   
36  26795140  t3_fyb9g  1299397232     0    pozhaluista   
50  29482904  t3_hjx5k  1306343697     0        visarun   
51  29608809  t3_hmmax  1306639850     0      [deleted]   
62  33263382  t3_jsy6u  1314200067     0  lsakbaetle3r9   

                                                title  \
35         Мам, а когда я выросту, ты постареешь?...    
36      Does the /r/ Ukraine have anyone here? Hello?   
50  Does immigration in Simferpol check whether th...   
51                     Tickets to Ukraine from Seoul?   
62                                looking for someone   

                                             selftext  score distinguish  \
35  Сегодня моя золотая птичка снова задалась вопр...      0         NaN   
36                                      Seems quiet.      10         NaN   
50  I have stayed 70 days and then 80 days out of ...      0         NaN   
51  Hi everyone.  I'd like to visit Ukraine for th...      2         NaN   
62  had a friend who moved to my town ~4-5 years a...      3         NaN   

    textlen  num_comments flair_text flair_css_class  
35      204             0        NaN             NaN  
36       13            24        NaN             NaN  
50      223             1        NaN             NaN  
51      281             0        NaN             NaN  
62      386             1        NaN             NaN

In [13]:
##remove the rows with non-enlish text
!pip install langdetect
from langdetect import detect
df['detect'] = detect(str(df['selftext']))

     |████████████████████████████████| 981 kB 4.5 MB/s eta 0:00:01
  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993242 sha256=d3497f5a98913e798bb0747b31419fc63e566e4f321b0edd58b3d902df4b6a30
  Stored in directory: /home/jovyan/.cache/pip/wheels/d1/c1/d9/7e068de779d863bc8f8fc9467d85e25cfe47fa5051fff1a1bb
Successfully built langdetect


In [14]:
df.head()

idint     idstr     created  nsfw         author  \
35  26781072  t3_fy0eo  1299351508     0         margys   
36  26795140  t3_fyb9g  1299397232     0    pozhaluista   
50  29482904  t3_hjx5k  1306343697     0        visarun   
51  29608809  t3_hmmax  1306639850     0      [deleted]   
62  33263382  t3_jsy6u  1314200067     0  lsakbaetle3r9   

                                                title  \
35         Мам, а когда я выросту, ты постареешь?...    
36      Does the /r/ Ukraine have anyone here? Hello?   
50  Does immigration in Simferpol check whether th...   
51                     Tickets to Ukraine from Seoul?   
62                                looking for someone   

                                             selftext  score distinguish  \
35  Сегодня моя золотая птичка снова задалась вопр...      0         NaN   
36                                      Seems quiet.      10         NaN   
50  I have stayed 70 days and then 80 days out of ...      0         NaN   
51  Hi everyone.  I'd like to visit Ukraine for th...      2         NaN   
62  had a friend who moved to my town ~4-5 years a...      3         NaN   

    textlen  num_comments flair_text flair_css_class detect  
35      204             0        NaN             NaN     en  
36       13            24        NaN             NaN     en  
50      223             1        NaN             NaN     en  
51      281             0        NaN             NaN     en  
62      386             1        NaN             NaN     en

In [15]:
df = df[df['detect'] == 'en']

In [16]:
df.shape

(16665, 14)

### Q. The command above did not give correct results e.g., the first row is in Russian, but the detect column says it is English. It is why the dataframe still have 16665 rows. So I tried the command below, but encountered error. I just moved on to next steps. We can apply the language detection codes later and run the notebook again:)


In [17]:
df['detect2'] = df['selftext'].apply(detect) 

LangDetectException: No features in text.

### 1-3. Cleaning Text Data

In [18]:
!pip install spacy
!python -m spacy download en_core_web_sm

2022-07-19 22:58:36.894343: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-07-19 22:58:36.894399: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [19]:
# Import spaCy
import spacy
# Load the English preprocessing pipeline
nlp = spacy.load('en_core_web_sm')

2022-07-19 22:59:46.586790: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-07-19 22:59:46.586845: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.


In [20]:
# Test: Parse the first reddit post in the dataset
parsed_post = nlp(df.selftext.iloc[0])
print(parsed_post)

Сегодня моя золотая птичка снова задалась вопросом... "Мам, а когда я выросту, ты постареешь?... а потом умрешь? ... и папа?... Мамочка, я не хочу взрослеть, можно я навсегда останусь такой маленькой?..."


In [21]:
# Print each sentence in the parsed post
for idx, sentence in enumerate(parsed_post.sents):  
    ##In python, .sents is used for "sentence segmentation" which is present inside spacy. 
    print(f'Sentence {idx + 1}')
    print(sentence)
    print('') #space

Sentence 1
Сегодня моя золотая птичка снова задалась вопросом...

Sentence 2
"Мам, а когда я выросту, ты постареешь?...

Sentence 3
а потом умрешь? ...

Sentence 4
и папа?...

Sentence 5
Мамочка, я не хочу взрослеть, можно я навсегда останусь такой маленькой?..."



In [22]:
# Extract the first 15 items for the following properties of the parsed post

# The token text 
token_text = [token.orth_ for token in parsed_post][:15]   
# Part of speech 
token_pos = [token.pos_ for token in parsed_post][:15]   
# Lemma (or 'dictionary form')
token_lemma = [token.lemma_ for token in parsed_post][:15]
# Stop word? t/f
token_stop = [token.is_stop for token in parsed_post][:15]
# Puncutation? t/f
token_punct = [token.is_punct for token in parsed_post][:15]

# Make a dataframe with these items
pd.DataFrame(zip(token_text, token_pos, token_lemma, token_stop, token_punct),
             columns=['token_text', 'part_of_speech', 'token_lemma', 'token_stop', 'token_punct'])

token_text part_of_speech token_lemma  token_stop  token_punct
0     Сегодня          PROPN     Сегодня       False        False
1         моя           NOUN         моя       False        False
2     золотая           VERB     золотая       False        False
3      птичка          PROPN      птичка       False        False
4       снова          PROPN       снова       False        False
5    задалась          PROPN    задалась       False        False
6    вопросом          PROPN    вопросом       False        False
7         ...          PUNCT         ...       False         True
8           "          PUNCT           "       False         True
9         Мам          PROPN         Мам       False        False
10          ,          PUNCT           ,       False         True
11          а          PROPN           а       False        False
12      когда          PROPN       когда       False        False
13          я          PROPN           я       False        False
14    выросту          PROPN     выросту       False        False

### 1-4. Preprocessing all data

In [23]:
import spacy
nlp = spacy.load('en_core_web_sm')
from gensim.models.phrases import Phrases, Phraser

In [24]:
def clean(token):
    """Helper function that specifies whether a token is:
        - punctuation
        - space
        - digit
    """
    return token.is_punct or token.is_space or token.is_digit

def line_read(df, text_col='selftext'):
    """Generator function to read in text from df and get rid of line breaks."""    
    for text in df[text_col]:
        yield text.replace('\n', '')

def preprocess(df, text_col='selftext', allowed_postags=['NOUN', 'ADJ']):
    """Preprocessing function to apply to a dataframe."""
    for parsed in nlp.pipe(line_read(df, text_col), batch_size=1000, disable=["tok2vec", "ner"]):
        # Gather lowercased, lemmatized tokens
        tokens = [token.lemma_.lower() if token.lemma_ != '-PRON-'
                  else token.lower_ 
                  for token in parsed if not clean(token)]
        # Remove specific lemmatizations, and words that are not nouns or adjectives
        tokens = [lemma
                  for lemma in tokens
                  if not lemma in ["'s",  "’s", "’"] and not lemma in allowed_postags]
        # Remove stop words
        tokens = [token for token in tokens if token not in spacy.lang.en.stop_words.STOP_WORDS]
        yield tokens

In [25]:
# This may take a while
lemmas = [line for line in preprocess(df)]

In [28]:
lemmas[2]

['stayed',
 'days',
 'days',
 'days',
 'know',
 'borispol',
 'pretty',
 'strict',
 'checking',
 'ones',
 'passport',
 'simferopol',
 'airport',
 'strict',
 'better',
 'use',
 'land',
 'border',
 'crossing',
 'lviv',
 'poland']

### 1-5.Phrase Modeling with `gensim`

In [30]:
from gensim.models.phrases import Phrases, Phraser

# Create bigram and trigram models
bigram = Phrases(lemmas, min_count=10, threshold=100)
trigram = Phrases(bigram[lemmas], min_count=10, threshold=50)  
bigram_phraser = Phraser(bigram)
trigram_phraser = Phraser(trigram)

# Form trigrams
trigrams = [trigram_phraser[bigram_phraser[doc]] for doc in lemmas]

In [31]:
# Join each into a string
trigrams_joined = [' '.join(trigram) for trigram in trigrams]
trigrams_joined[0]

'сегодня моя золотая птичка снова задалась вопросом мам а когда я выросту ты постареешь а потом умрешь и папа мамочка я_не хочу взрослеть можно я навсегда останусь такой маленькой'

In [32]:
#We can use .keys() to identify the bigrams in the dataset. How many bigrams were identified by the parser?
len(bigram_phraser.phrasegrams.keys())

542

In [33]:
#Look at biagrams
list(bigram_phraser.phrasegrams.keys())[:10]

['border_crossing',
 'shed_light',
 'years_ago',
 'hey_guys',
 'thanks_advance',
 'eastern_europe',
 'eastern_european',
 'english_speakers',
 'cross_border',
 'story_short']

In [34]:
# Look at trigrams
[trigram for trigram in list(trigram_phraser.phrasegrams.keys()) if trigram.count('_') == 2]

['long_story_short',
 'ministry_foreign_affairs',
 'euromaidan_pr_twitter](http://goo.gl',
 'euromaidan_pr_facebook](http://goo.gl',
 'ukrainian_brothers_sisters',
 'minister_foreign_affairs',
 'having_trouble_finding',
 'un_security_council',
 'worst_case_scenario',
 'united_states_america',
 'native_english_speaker',
 '=_webp&s_=',
 'black_sea_fleet',
 'view_poll](https://www.reddit.com_poll',
 'якщо_у_вас',
 'sovereignty_territorial_integrity',
 '=_share&utm_medium',
 'ukraine_🇺_🇦',
 'anti_tank_weapons',
 'donetsk_luhansk_regions',
 'committing_war_crimes',
 'news_spread_likely',
 'russia_backed_trolls',
 'reception_points_ready',
 "right_edit_don't",
 'need_visa_pass',
 'need_passport_visas',
 "suspended_don't_need",
 'need_visa:•_\u2060in',
 'ukrainian_https://www.gov.pl/web/udsc/ukraina---ua_\u2060in',
 'copy_paste_encourage',
 'https://www.gov.pl/web/udsc/ukraina---ua_\u2060in_english',
 'використовував_google_translate',
 'bic_nbua_ua',
 'address_instytutska_st',
 'initial_eval

### 1-6. Save the file after preprocessing

In [35]:
# Inserting next to selftext column
df.insert(loc=7, column='lemmas', value=trigrams_joined)
# Removing empty rows in lemmas
df = df[~df['lemmas'].isin([''])]

In [36]:
df.head()

idint     idstr     created  nsfw         author  \
35  26781072  t3_fy0eo  1299351508     0         margys   
36  26795140  t3_fyb9g  1299397232     0    pozhaluista   
50  29482904  t3_hjx5k  1306343697     0        visarun   
51  29608809  t3_hmmax  1306639850     0      [deleted]   
62  33263382  t3_jsy6u  1314200067     0  lsakbaetle3r9   

                                                title  \
35         Мам, а когда я выросту, ты постареешь?...    
36      Does the /r/ Ukraine have anyone here? Hello?   
50  Does immigration in Simferpol check whether th...   
51                     Tickets to Ukraine from Seoul?   
62                                looking for someone   

                                             selftext  \
35  Сегодня моя золотая птичка снова задалась вопр...   
36                                      Seems quiet.    
50  I have stayed 70 days and then 80 days out of ...   
51  Hi everyone.  I'd like to visit Ukraine for th...   
62  had a friend who moved to my town ~4-5 years a...   

                                               lemmas  score distinguish  \
35  сегодня моя золотая птичка снова задалась вопр...      0         NaN   
36                                              quiet     10         NaN   
50  stayed days days days know borispol pretty str...      0         NaN   
51  hi like visit ukraine time looking cheap ticke...      2         NaN   
62  friend moved town ~4 years ago.he myspace acce...      3         NaN   

    textlen  num_comments flair_text flair_css_class detect  
35      204             0        NaN             NaN     en  
36       13            24        NaN             NaN     en  
50      223             1        NaN             NaN     en  
51      281             0        NaN             NaN     en  
62      386             1        NaN             NaN     en

In [37]:
# Save to new csv
df.to_csv('ukraine_lemmas.csv', index=False)

In [38]:
 df.head()

idint     idstr     created  nsfw         author  \
35  26781072  t3_fy0eo  1299351508     0         margys   
36  26795140  t3_fyb9g  1299397232     0    pozhaluista   
50  29482904  t3_hjx5k  1306343697     0        visarun   
51  29608809  t3_hmmax  1306639850     0      [deleted]   
62  33263382  t3_jsy6u  1314200067     0  lsakbaetle3r9   

                                                title  \
35         Мам, а когда я выросту, ты постареешь?...    
36      Does the /r/ Ukraine have anyone here? Hello?   
50  Does immigration in Simferpol check whether th...   
51                     Tickets to Ukraine from Seoul?   
62                                looking for someone   

                                             selftext  \
35  Сегодня моя золотая птичка снова задалась вопр...   
36                                      Seems quiet.    
50  I have stayed 70 days and then 80 days out of ...   
51  Hi everyone.  I'd like to visit Ukraine for th...   
62  had a friend who moved to my town ~4-5 years a...   

                                               lemmas  score distinguish  \
35  сегодня моя золотая птичка снова задалась вопр...      0         NaN   
36                                              quiet     10         NaN   
50  stayed days days days know borispol pretty str...      0         NaN   
51  hi like visit ukraine time looking cheap ticke...      2         NaN   
62  friend moved town ~4 years ago.he myspace acce...      3         NaN   

    textlen  num_comments flair_text flair_css_class detect  
35      204             0        NaN             NaN     en  
36       13            24        NaN             NaN     en  
50      223             1        NaN             NaN     en  
51      281             0        NaN             NaN     en  
62      386             1        NaN             NaN     en

### For the next steps, make sure to use the file: 'ukraine_lemmas.csv'